src/consul.rs

```diff
@@ -0,0 +1,308 @@
+use anyhow::Result;
+use base64::prelude::{Engine as _, BASE64_STANDARD};
+use reqwest::{StatusCode, Url};
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use std::{
+    collections::HashMap,
+    time::{Duration, Instant, SystemTime},
+};
+use tokio::time::sleep;
+
+use crate::dns_trait::DnsType;
+
+/// A DNS record based on the tags of a service in Consul
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Hash)]
+pub struct DnsRecord {
+    pub hostname: String,
+    #[serde(rename = "type")]
+    pub type_: DnsType,
+    pub ttl: Option<i32>,
+    pub value: String,
+}
+
+#[derive(Debug, Eq, PartialEq, Hash, Serialize, Deserialize)]
+struct ConsulLock {
+    pub locked_at: SystemTime,
+}
+
+impl ConsulLock {
+    fn new() -> Self {
+        Self {
+            locked_at: SystemTime::now(),
+        }
+    }
+}
+
+#[derive(Debug, Deserialize)]
+struct ConsulKVResponse {
+    #[serde(rename = "Value")]
+    value: Option<String>,
+}
+
+pub struct ConsulClient {
+    pub http_client: reqwest::Client,
+    pub kv_api_base_url: Url,
+    pub catalog_api_base_url: Url,
+    pub datacenter: Option<String>,
+}
+
+impl ConsulClient {
+    pub fn new(consul_address: Url, consul_datacenter: Option<String>) -> Result<ConsulClient> {
+        let kv_api_base_url = consul_address.join("v1/")?.join("kv/")?;
+        let catalog_api_base_url = consul_address.join("v1/")?.join("catalog/")?;
+        let client = reqwest::Client::builder().build()?;
+        Ok(ConsulClient {
+            http_client: client,
+            kv_api_base_url,
+            catalog_api_base_url,
+            datacenter: consul_datacenter,
+        })
+    }
+
+    /// Acquire a lock
+    ///
+    /// Times out after a while and returns an error if it does.
+    pub async fn acquire_lock(&self) -> Result<()> {
+        let consul_lock = ConsulLock::new();
+        let wait_time = Instant::now();
+        let timeout = Duration::from_secs(10);
+
+        loop {
+            if wait_time.elapsed() > timeout {
+                println!("Timed out trying to acquire lock");
+                println!("Assuming poisoned lock, deleting last lock");
+                self.drop_lock().await?;
+            }
```

Comment:
This locking method isn't safe, especially now that we're properly blocking on changes to the Consul state, since we have no idea if some other instance thinks we timed out while we're still actively changing the state and records.  We should use the actual Consul locking support instead.

Unfortunately that is slightly more involved as it involves creating and renewing sessions that you then acquire the lock within.  I'll send you an example of how to use it on Discord.  Let me know if you'd like to go through how it works together.

All of these operations should also be performed against the DC specified on the command line, if one was passed in.  Otherwise we'd be using the same lock for different DCs.